### ISSUE TITLE
attribute editor: autopopulate attribute info from CSV and NetCDF

### ISSUE THREAD

**Issue by mbjones** [NCEAS/metacatui#202]
Use info from CSV files and NetCDF files to autopopulate the attribute metadata.  In particular:

For CSV files:
- [ ] Parse the column header to get the column names and number of header lines to skip
- [ ] Parse the data to:
    - [ ] Guess the type of each column
    - [ ] For enumerated columns, find the unique values and pre-populate enumerated codes, awaiting a definition

For NetCDF files:
- [ ] pull existing attribute definitions from netcdf and populate the right fields

**Comment by mbjones** [comment]
Also consider adding, possibly in a phase 2:

- [ ] parse CSVW (CSV for the Web) metadata formats like package.json for entity and attribute metadata

**Comment by amoeba** [comment]
This came up again on our salmantics call yesterday and I wanted to see how feasible it'd be to at least form the basis of this feature and implement just

> [ ] Parse the column header to get the column names and number of header lines to skip

I think this feature alone is probably costing our team and submitters a lot of time so I think getting something integrated soon would be really great.

I took a shot at implementing it today and I'm pretty happy with the result. I opted not to pull in a full CSV library for a few good reasons and instead wrote a pretty simple routine that reads just the start of a file and tries to find the header names. I tested it with the CSV test suites from `readr` and `data.table` and came up with two limitations (1) my approach does not like the presence of a UTF BOM and (2) my approach doesn't handle quoted column names with commas inside them. Both of these aren't common IME.

A demo is available at https://neutral-cat.nceas.ucsb.edu. Create a new package, add a CSV, hit Describe, and you should find the Fill... button.

![2021-02-12 12 52 48](https://user-images.githubusercontent.com/563/107826436-5e6da080-6d31-11eb-8d39-b3114fb7c4f3.gif)

This approach will overwrite attribute names on attributes but leave them otherwise in-tact. It will also create new, blank attributes if needed. I think this approach will work well as-is and we could choose to extend it to support Excel, NetCDF, TSV, etc at later dates.

I'd love feedback of really any sort here. Here's the [diff](https://github.com/NCEAS/metacatui/compare/develop...feature-autofill-csv?expand=1) if anyone wants to check that out. And, again, the demo is at https://neutral-cat.nceas.ucsb.edu.

**Comment by laijasmine** [comment]
This sounds awesome @amoeba .. but I can't seem to add files to new submissions or edit existing datasets on [neural-cat](https://neutral-cat.nceas.ucsb.edu/). I remember being able to before when testing out annotations for #1402 So maybe this is a permissions issue? Happy to pass this along to interns as well to test once we sort this out.

**Comment by jeanetteclark** [comment]
I was able to submit and edit a dataset without issue.

I did run into one bug. I was trying out replacing a file and re-loading the attributes from the new file. All of the information was retained, but all of the attributes showed up with a red flag that the information was complete as soon as I clicked the button. When I click on the attribute name in the left hand menu, it turns green. Other than that minor issue, this looks great so far! I haven't done any testing with different/weirdly formatted csvs or netcdfs though.

Before clicking "fill from file" :
<img width="1097" alt="Screen Shot 2021-02-12 at 3 11 37 PM" src="https://user-images.githubusercontent.com/24237074/107827888-e4dfad80-6d44-11eb-9e3b-862c4d301d59.png">

After:
<img width="1089" alt="Screen Shot 2021-02-12 at 3 11 44 PM" src="https://user-images.githubusercontent.com/24237074/107827929-fd4fc800-6d44-11eb-9136-f5077aef2746.png">

**Comment by laijasmine** [comment]
thanks Bryce - fixed after clearing my cache! 

No glaring issues in my tests. Also sent it to the interns to test

**Comment by BrennieDev** [comment]
Attempting to test this out by making a dataset in the UI and adding/describing a csv, I do not see the button popping up. It seems to be specifally for this data set, as I can see the button in one that @laijasmine uploaded.

This is the uncooperative data set: https://neutral-cat.nceas.ucsb.edu/view/urn:uuid:c9e2e029-6fa8-4917-ba55-2410f400f493

**Comment by laijasmine** [comment]
Found a scenario with @BrennieDev that might be a bug:
1. create a new dataset
2. upload a data file
3. save your dataset
4. go back and edit the same dataset again
5. click describe for the datafile you uploaded earlier
5. click fill from file
Does not add new attributes. Sometimes users with save frequently to make sure they don't lose their changes so I can see this being an issue for some users.

**Comment by amoeba** [comment]
Yeah, nice catch @laijasmine and @BrennieDev. I thought about that early on and forgot to come back to it.

What's going on is that, when you come back to a package later, we've lost the `File` handle that we originally obtained from picking the file from our computer. We _could_ get that back by having the user find the file again on their filesystem but this is likely frustrating. I've since updated the routine so it will fetch the object from the server if it needs to and then it'll work with that. How well this works for the user will depend on their internet connection and how large the file(s) are so I'm tempted to put a warning or hard limit on this.

@jeanetteclark thanks for testing so quickly. That looks pretty funky and I think it should be addressed. I hooked into the existing validation routines so it looks like this might be an edge case in those. I'll take a look.

**Comment by amoeba** [comment]
Hey @jeanetteclark, I tweaked things a bit and hopefully the behavior is better. Validation for each attribute should now work like before in that it only shows when you click on an item in the sidebar or change items. If you have time, it'd be great if you could pop back over to https://neutral-cat.nceas.ucsb.edu and check it out again.

**Comment by amoeba** [comment]
Hey @BrennieDev, you caught a weird edge case not entirely related to this feature. For some reason, the editor just generally thinks that's an Excel file. The only odd thing about it is that it has a UTF-8 BOM sequence (\uEFBBBF). I should be able to get it fixed.

**Comment by yvanlebras** [comment]
It appears to me this is a very important functionality! Thank you for working on it! Have a nice week!

**Comment by amoeba** [comment]
We reviewed this on today's developer and there was agreement that this might not be exactly what we ship to our production repositories yet but that it would likely be very useful for the data team. I'm going to feature flag it so we can turn it on over on test.arcticdata.io and get the team using it while we work on the first public version of this.

Things we might look into before a first public release are:

1. Auto-fill automatically when creating a new entity rather than requiring a click
2. Add a way to get yourself out of a situation where you auto-fill too many attributes to fill in in one go so the use can save (e.g., clear auto-filled attributes that otherwise haven't been edited)
3. Expanding to a more robust CSV parser (such as PapaParse or SheetJS) and possibly adding support for other formats such as NetCDF

**Comment by laijasmine** [comment]
I'll try to see what cases would be appropriate so that we can get some feedback for you @amoeba.

Right now we use `EML::shiny_attributes` that helps us grab the attribute names. This feature might not get as much use because we rarely create new datasets from scratch on test while most of the time datasets start on production. I want to avoid getting interns to switch back and forth using test and production might cause some confusion.

**Comment by amoeba** [comment]
Thanks @laijasmine, that makes sense. 

Maybe that's a good reason to put a hold on my feature branch here (but not this ticket, generally) while we figure out what we do want to ship.

**Comment by laurenwalker** [comment]
Did some testing and a bit of cleanup and this MVP is good to go for MetacatUI 2.15.0 (merged to develop)

**Comment by amoeba** [comment]
Thanks @laurenwalker, those look like good improvements.

**Comment by jeanetteclark** [comment]
can we close this? it doesn't implement everything described in the issue, but maybe those items can be split into a separate one if needed?

**Comment by mbjones** [comment]
Yes. Can we check off/summarize what was done, and summarize what remains. Opening another issue with those items would be helpful if we think they are something we'd like to tackle at some point.

**Comment by robyngit** [comment]
I've opened new issues for the remaining tasks discussed here:
- https://github.com/NCEAS/metacatui/issues/2055
- https://github.com/NCEAS/metacatui/issues/2056
- https://github.com/NCEAS/metacatui/issues/2057
- https://github.com/NCEAS/metacatui/issues/2058
- https://github.com/NCEAS/metacatui/issues/2059
- https://github.com/NCEAS/metacatui/issues/2060
- https://github.com/NCEAS/metacatui/issues/2061
